### PR TITLE
Markdown Toolbar Actions Implementation

### PR DESCRIPTION
**Pull request by chlafreniere**
Looking for feedback here for code design + sanity checking here, so that's why I'm sending my PR out early; if anyone strongly disagrees with anything here, I just want to have enough time to react 😄.

Yes, I will be adding tests for the markdown transformer. Tomorrow 😛.

This is the "crawl" for the markdown toolbar implementation. Currently does not support undoing text that's already been transformed (i.e. if you click the bold button twice, it does not un-bold the text). This can be added pretty easily, but I want to avoid scope creep in this initial PR.

This code is based off of the latest branch that @halerankin is working off of. Since he has the frontend bits there, I wanted to keep my changes separate, so that's why this PR is being compared to his branch. When we can check-in his bits then we can create a separate PR into master... Or @halerankin can take these changes into his branch and address PR comments there... Honestly, I'm ok with either option, so we can sync.

Note: I did change how things work slightly and now have an actionbar instead of having the toolbar be defined in HTML. There are a few reasons for this, including accessibility, ability to contribute action bar items via extensions (TODO!), and overall consistency with the rest of the app to use. I had to tweak some css values because of this; @halerankin I'll bug you about how to make this pixel perfect again! Sorry about that.

### CODE REVIEW

File: src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts [hunk 1]
```diff
@@ -24,6 +29,47 @@ export class MarkdownToolbarComponent {
 	public buttonImage = localize('buttonImage', "Image");
 	public buttonPreview = localize('buttonPreview', "Markdown preview toggle - off");
 
-	constructor() {
+	public get cellModel(): ICellModel {
+		return this._cellModel;
+	}
+
+	@Input() public set cellModel(value: ICellModel) {
+		this._cellModel = value;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts:37]: Why both having separate gettings/setters/backing value when you could just put make the property public? That should still work with the @Input decorator.

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+		tooltip: string,
+		private _cellModel: ICellModel,
+		private _type: MarkdownButtonType,
+		@INotebookService private _notebookService: INotebookService
+	) {
+		super(id, label, cssClass);
+		this._tooltip = tooltip;
+	}
+	public run(context: any): Promise<boolean> {
+		return new Promise<boolean>((resolve, reject) => {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:35]: Would be great to just use async/await for these

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+			// -1 is considered as having no explicit offset, so do not do anything with selection
+			default: return -1;
+		}
+	}
+
+	private getEditorControl(): CodeEditorWidget {
+		if (!this._notebookEditor) {
+			this._notebookEditor = this._notebookService.findNotebookEditor(this._cellModel.notebookModel.notebookUri);
+		}
+		if (this._notebookEditor && this._notebookEditor.cellEditors && this._notebookEditor.cellEditors.length > 0) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:188]: nit: You can shorten this by just doing `this._notebookEditor?.cellEditors?.length > 0`

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+			}
+		}
+		return undefined;
+	}
+
+	private editorHasNoSelection(selection: Selection): boolean {
+		return !selection || (selection.startLineNumber === selection.endLineNumber && selection.startColumn === selection.endColumn);
+	}
+
+	// Set selection (which also controls the cursor) for a given button type
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:206]: Use actual XML comments for these

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+	private setEndSelection(range: IRange, type: MarkdownButtonType, editorControl: CodeEditorWidget, noSelection: boolean) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:207]: Specify return types

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+			let newRange: IRange = {
+				startColumn: range.startColumn + offset,
+				endColumn: range.endColumn + offset,
+				startLineNumber: range.startLineNumber,
+				endLineNumber: range.endLineNumber
+			};
+			editorControl.setSelection(newRange);
+		}
+		// Always give focus back to the editor after pressing the button
+		editorControl.focus();
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:222]: Probably want to set this if range is undefined at the top too, right?

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+			case MarkdownButtonType.LINK:
+				return 2;
+			case MarkdownButtonType.IMAGE:
+				return 0;
+			// -1 is considered as having no explicit offset, so do not do anything with selection
+			default: return -1;
+		}
+	}
+
+	private getEditorControl(): CodeEditorWidget {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:184]: Return type `CodeEditorWidget | undefined`

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+				markdownTextTransformer.transformText(this._type);
+				resolve(true);
+			} catch (e) {
+				reject(e);
+			}
+		});
+	}
+}
+
+export class MarkdownTextTransformer {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:47]: Does this need to be exported?

File: src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts (hunk 3 above)
Comment (chlafreniere) [reply src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:188]: I keep forgetting this syntax exists now. Thank you. Definitely much cleaner.

Comment (halerankin) [reply src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts:188]: That's some sweet implied null checking there!

**Pull request merged**